### ISSUE TITLE
oauth2 pass state parameter to server

### ISSUE THREAD

**Issue by yaniv51** [nuxt-community/auth-module#688]
### What problem does this feature solve?

Getting state information while user request to login with oauth2.

While pass a state for oauth2 login, the value is passed to provider(google,facebook) and the provider is sending this value back to callback URL but oauth2 implementation is ignoring this state paramter.

It is possible to pass this parameter via the POST request

<div align="right"><sub><em>This feature request is available on <a href="https://cmty.app/nuxt">Nuxt</a> community (<a href="https://cmty.app/nuxt/auth-module/issues/c577">#c577</a>)</em></sub></div>

**Comment by kevintechie** [comment]
Are you saying the Google and Facebook are passing the state parameter via POST back to the client? While my experience is with AWS Cognito, it is my understanding that the OAuth2 sepc says that the state parameter should be passed back to the client on the query string.

I can confirm that nuxt-auth correctly picks up the state parameter on the query string when using the Oauth2 scheme.

**Comment by yaniv51** [comment]
No, when sending state parametr to Google/Facebook, they pass it back as a parameter in the redirect URL(and as you said, nuxt-auth is implementing this usage and works perfect)

All I ask is to pass this parameter to server while sending POST request to token endpoint.
Currently the state is only for NUXT validation and I want it also to be validated via my backend (ouaht2.ts -> _handleCallback  -> line 277)

Today I'm using the referer header to validate it but it could be much easier

**Comment by kevintechie** [comment]
It's likely what you're asking for would not be Oauth2 compliant and the token endpoint would reject the token request if the state parameter was in a the POST request to the endpoint. The purpose of the state parameter is to help prevent CSRF attacks by comparing the state sent to the authentication endpoint to the state returned by the authentication endpoint so that a fraudulent response will be rejected if the state does not match. The authentication server will return a different code (the authorization code) which you send to the token endpoint along with other information. From there, the token endpoint will send you a signed token you can use with your API endpoints. What Oauth2 provider are you using that requires state to be passed to it?

**Comment by yaniv51** [comment]
I'm using it for a different purpose and not for the CSRF attack.

In my case, when user login he is asking to accept terms.
After user accept multiple terms(not all of them must be checked) and I make login with google - I would like to get terms status to backend (for save this data)
Because AUTH module is sending the request - I can not change the data on POST request and the only bypass I thought is to use the state object.

It might be that a new feature is required - pass a new object that would saved on local store:
$auth.loginWith("google", {}, "SOMEDATA")
And while request to login on callback page(ouaht2.ts -> _handleCallback ) - send this object to server.

I can implement it alone and to save this state before login and use it after - but I think it could be useful to pass data to backend on login request.

It is not an issue on local login since it is possible to control which object is sent to backend...

**Comment by kevintechie** [comment]
I think what you’ll have to do is to save the info you want to send to the server to a cookie using the state as the key. Then when redirecting back, read the cookie data on the server.
